frappe/public/js/legacy/form.js

```diff
@@ -922,6 +922,17 @@ _f.Frm.prototype.action_perm_type_map = {
 
 _f.Frm.prototype.validate_form_action = function(action) {
 	var perm_to_check = this.action_perm_type_map[action];
+	var allowed_for_workflow = false;
+
+	// Allow submit, write and create permissions for read only documents that are assigned by 
+	// workflows if the user already have those permissions. This is to allow for users to 
+	// continue through the workflow states and to allow execution of functions like Duplicate.
+	if ((frappe.workflow.is_read_only(this.doctype, this.docname) && 
+		(frappe.perm.get_perm(this.doc.doctype)[0].write && perm_to_check == "write")) ||
```

Comment:
set `perms = frappe.perm.get_perm(this.doc.doctype)[0]` so you call the function only once!